### DISCUSSION TITLE
OpenGL sample does not work on Windows 10 (works on MacOS)

### DISCUSSION THREAD

**Discussion by ArtjomP** [AvaloniaUI/Avalonia#6396] [Q&A]
Hello there!
I'm trying to run the OpenGL rendering sample.
I created [a sample project](https://github.com/ArtjomP/AvaloniaOpenGlTest) based on [OpenGlPage](https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/OpenGlPage.xaml). I start it and it does not show the teapot only blank space. Also the break point never stops inside the the 
```cs
OnOpenGlRender(GlInterface gl, int fb)
```
![image](https://user-images.githubusercontent.com/20728465/128822084-52a6980f-9d59-4db0-b939-e8db049f3e0f.png)

Why the sample doesn't work on Windows?
How can I check in code that OpenGL rendering is not available?
Thank you in advance!

**Comment by kekekeks** [comment]
**TL;DR:** either add `.With(new Win32PlatformOptions { UseWgl = true})` or wait till September.


On Windows we are using ANGLE for rendering. ANGLE is a library from Google that implements OpenGL ES on top of DirectX. We need this because on Windows it's somehow way more stable than using native OpenGL (i. e. native OpenGL might be dysfunctional or outright missing with Microsoft-provided drivers. 

Unfortunately ANGLE doesn't support multithreading. It doesn't really support shared OpenGL contexts either (all contexts inside one EGLDisplay instance are backed by a single D3DDevice). So we can't safely use the context sharing feature to provide OpenGL API to the user.

We do, however, have experimental support for native OpenGL on windows, which can be turned on via UseWgl platform option. When running in this mode, context sharing is enabled and supported.


To properly solve this issue, I'm currently working on safely passing render results from native OpenGL to ANGLE via DXGI shared handles, but it's not completely ready yet and is not guaranteed to be included in the stable 0.10.x branch till september.

**Reply by kekekeks** [reply]
You can try using a build from this PR though - https://github.com/AvaloniaUI/Avalonia/pull/6391
